src/library/scala/collection/immutable/HashMap.scala

```diff
@@ -168,23 +167,53 @@ final class HashMap[K, +V] private[immutable] (private[immutable] val rootNode:
     else new HashMap(transformed)
   }
 
-  override def filterImpl(pred: ((K, V)) => Boolean, flipped: Boolean): HashMap[K, V] = {
-    // This method has been preemptively overridden in order to ensure that an optimizing implementation may be included
-    // in a minor release without breaking binary compatibility.
-    //
-    // In particular, `filterImpl` could be optimized to filter and reconstruct the trie node-by-node, without having to
-    // perform any hashing or equality checks.
-    super.filterImpl(pred, flipped)
+  override protected[collection] def filterImpl(pred: ((K, V)) => Boolean, isFlipped: Boolean): HashMap[K, V] = {
+    val newRootNode = rootNode.filterImpl(pred, isFlipped)
+    if (newRootNode eq rootNode) this
+    else if (newRootNode.size == 0) HashMap.empty
+    else new HashMap(newRootNode)
   }
 
   override def removeAll(keys: IterableOnce[K]): HashMap[K, V] = {
-    // This method has been preemptively overridden in order to ensure that an optimizing implementation may be included
-    // in a minor release without breaking binary compatibility.
-    //
-    // In particular, `removeAll` could be optimized to avoid reallocating the `HashMap` wrapper of the rootNode on each
-    // element in `keys`, and potentially to take advantage of the structure of `keys`, if it happens to be a HashSet
-    // which would allow us to skip hashing keys all together.
-    super.removeAll(keys)
+    if (isEmpty) {
+      this
+    } else {
+      keys match {
+        case hashSet: HashSet[K] =>
+          if (hashSet.isEmpty) {
+            this
+          } else {
+            var curr = rootNode
+            new ChampBaseIterator(hashSet.rootNode) {
+              while(curr.size > 0 && hasNext) {
```

Comment:
Is there a reason that this loop is done in the constructor of the iterator? `curr` gets boxed on the heap (in an `s.r.ObjectRef`) because the constructor doesn't have access to `removeAll`'s stack

Reply:
Resolved this by extracting the logic out into a `class MapNodeRemoveAllSetNodeIterator[A]`